internal/expr/prepare.go

```diff
@@ -199,18 +234,96 @@ func prepareOutput(ti typeNameToInfo, p *outputPart) ([]fullName, []typeMember,
 		for _, c := range p.source {
 			outCols = append(outCols, c)
 		}
-		return outCols, typeMembers, nil
+		return &outputChunk{outCols}, typeMembers, nil
 	}
 
 	// Case 2.2: No columns e.g. "(&P.name, &P.id)".
 	for _, t := range p.target {
 		outCols = append(outCols, fullName{name: t.name})
 	}
-	return outCols, typeMembers, nil
+	return &outputChunk{outCols}, typeMembers, nil
 }
 
 type typeNameToInfo map[string]typeInfo
 
+type sqlChunk interface {
+	chunk()
+}
+
+type outputChunk struct {
+	outCols []fullName
+}
+
+func (outputChunk) chunk() {}
+
+type inputChunk struct{}
+
+func (inputChunk) chunk() {}
+
+type inChunk struct {
+	typeMembers []typeMember
+}
+
+func (inChunk) chunk() {}
+
+type bypassChunk struct {
+	str string
+}
+
+func (bypassChunk) chunk() {}
+
+func generateInputSQL(inCount int) (int, string) {
+	return inCount + 1, "@sqlair_" + strconv.Itoa(inCount)
+}
+
+func generateOutputSQL(outCount int, outCols []fullName) (int, string) {
+	var sql bytes.Buffer
+	for i, c := range outCols {
+		sql.WriteString(c.String())
+		sql.WriteString(" AS ")
+		sql.WriteString(markerName(outCount))
+		if i != len(outCols)-1 {
+			sql.WriteString(", ")
+		}
+		outCount++
+	}
+	return outCount, sql.String()
+}
+
+func generateInSQL(inCount int, typeMembers []typeMember) (int, string) {
+	var sql bytes.Buffer
+	sql.WriteString("IN (")
+	for i, tm := range typeMembers {
+		switch tm := tm.(type) {
+		case *structField:
+			sql.WriteString("@sqlair_")
+			sql.WriteString(strconv.Itoa(inCount))
+			if i < len(typeMembers)-1 {
+				sql.WriteString(", ")
+			}
+			inCount++
+		case *mapKey:
+			length := 1
+			if tm.slice != nil {
+				log.Printf("length is %d", tm.slice.length)
+				length = tm.slice.length
+			}
+			for j := 0; j < length; j++ {
+				sql.WriteString("@sqlair_")
+				sql.WriteString(strconv.Itoa(inCount))
+				if i < len(typeMembers)-1 || j < length-1 {
```

Comment:
This guard seems to have two responsibilities. Perhaps it would be cleaner if there were `if` here which adds a comma between map elements (`j < length - 1`), then a separate `if` at the end of the loop which adds a comma between type members (`i < len(typeMembers)-1`) at the end of the loop. This would also remove the need to add a comma in the `*structField` case

Reply:
Changed.